third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/TMEMAlloc1D.cpp

```diff
@@ -0,0 +1,229 @@
+namespace mlir {
+
+// Wrapper class to hold the context for handling
+// 1D TMEM Allocation.
+class TMEM1DAllocator {
+private:
+  OpBuilder &builder;
+  Block *globalStart;
+  int numWarps;
+  // Intermediate info to minimize code reuse across functions.
+  tt::ExpandDimsOp expandedInput = nullptr;
+  ttng::TMEMAllocOp allocOp = nullptr;
+
+public:
+  TMEM1DAllocator(OpBuilder &builder, Block *globalStart)
+      : builder(builder), globalStart(globalStart),
+        numWarps(ttg::lookupNumWarps(&globalStart->front())) {}
+
+private:
+  void copyAttrs(Operation *oldOp, Operation *newOp) {
+    // Right now we just copy over ttg.partition per
+    // the example.
+    // TODO: Should we copy over loop information?
+    auto partitionAttr = oldOp->getAttr("ttg.partition");
+    newOp->setAttr("ttg.partition", partitionAttr);
+  }
+
+  void setExpandedInput(tt::ExpandDimsOp expandedInput) {
+    this->expandedInput = expandedInput;
+  }
+
+  tt::ExpandDimsOp getExpandedInput() {
+    assert(expandedInput != nullptr && "Must call setExpandedInput");
+    return expandedInput;
+  }
+
+  void setAllocOp(ttng::TMEMAllocOp allocOp) { this->allocOp = allocOp; }
+
+  ttng::TMEMAllocOp getAllocOp() {
+    assert(allocOp != nullptr && "Must call getAllocOp()");
+    return allocOp;
+  }
+
+  RankedTensorType getResultTensorType(Operation *op, size_t expectedSize) {
+    auto outputType = dyn_cast<RankedTensorType>(op->getResult(0).getType());
+    if (!outputType || outputType.getShape().size() != 2) {
+      assert("Invalid tensor input");
+    }
+    return outputType;
+  }
+
+  ttng::TMEMAllocOp alloc1DTMEMBuffer() {
+    auto expandedInput = getExpandedInput();
+    auto oldRetType = getResultTensorType(expandedInput, 2);
+    SmallVector<int64_t> shape = {oldRetType.getShape().begin(),
+                                  oldRetType.getShape().end()};
+    auto context = builder.getContext();
+    auto oldEncoding = oldRetType.getEncoding();
+    Attribute tensorMemorySpace =
+        ttng::TensorMemorySpaceAttr::get(builder.getContext());
+    // TODO(njriasan): Do we need to handle the ScaleDotElemType::E2M1 && transA
+    // case at all from TCGen5MMAScaledOp::getBlockM?
+    auto blockM = shape[0];
+    auto elemType = oldRetType.getElementType();
+    unsigned elemBitWidth = elemType.getIntOrFloatBitWidth();
+    assert((elemBitWidth == 16 || elemBitWidth == 32) &&
+           "TMEM Layout don't support fp8");
+    bool unpacked = elemBitWidth != 16;
+    ArrayRef<unsigned> CTASplitNum =
+        ttg::getCTALayout(oldEncoding).getCTASplitNum();
+    auto encoding = ttng::TensorMemoryEncodingAttr::get(
+        builder.getContext(), blockM, shape[1],
+        /*unpacked=*/unpacked, CTASplitNum[0], CTASplitNum[1]);
+    auto tmemDesc =
+        ttg::MemDescType::get(shape, elemType, encoding, tensorMemorySpace,
+                              /*mutableMemory=*/true);
+
+    builder.setInsertionPointToStart(globalStart);
```

Comment:
Do we need to insert this at the top of the function as opposed to at the first user? Do we have anything that would hoist it out of the loop if we did the latter?